### PR TITLE
Fix redirect failures from apache client 5.x

### PR DESCRIPTION
**Pull request by corneil**
Provided alternative by using Reactor Netty HttpClient that does provide proper redirect handling.

Reason for redirect handler was in deleted `DropAuthorizationHeaderRequestRedirectStrategy`

### CODE REVIEW

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java [hunk 1]
```diff
@@ -153,67 +143,81 @@ public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolea
...
-	private HttpClientBuilder httpClientBuilder(SSLContext sslContext) {
-		// Register http/s connection factories
-		Lookup<ConnectionSocketFactory> connSocketFactoryLookup = RegistryBuilder.<ConnectionSocketFactory> create()
-			.register("https", new SSLConnectionSocketFactory(sslContext, NoopHostnameVerifier.INSTANCE))
-			.register("http", new PlainConnectionSocketFactory())
-			.build();
-		return HttpClients.custom()
-			.setConnectionManager(new BasicHttpClientConnectionManager(connSocketFactoryLookup));
+
+	private static void removeAuthorization(HttpHeaders headers) {
```
Comment (onobc) [inline spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:200]: This impl is missing all of the variants described in this methods javadocs ([above](https://github.com/spring-cloud/spring-cloud-dataflow/pull/5996/files#diff-d659fdfefb699e093be40defb784f08e85441f3a595328ee597aa899184ef6b4R152-R174)) which is the meat/potatoes of the deleted request strategy `DropAuthorizationHeaderRequestRedirectStrategy`.

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java [hunk 2]
```diff
@@ -153,67 +143,81 @@ public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolea
...
 		}
-
-		HttpComponentsClientHttpRequestFactory customRequestFactory =
-				new HttpComponentsClientHttpRequestFactory(
-						clientBuilder
-								.setRedirectStrategy(new DropAuthorizationHeaderRequestRedirectStrategy(extra))
-								// Azure redirects may contain double slashes and on default those are normilised
-								.setDefaultRequestConfig(RequestConfig.custom().build())
-								.build());
+		// TODO what do we do with extra?
```
Comment (onobc) [inline spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:219]: We use it an implemented the decision to drop or not drop the auth headers as the previous solution did.

**Review by onobc — CHANGES_REQUESTED** [review]
I like the switch to use Netty client but I think we are missing a huge chunk of logic that used to reside in the drop header strategy. Am I missing something?

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java (hunk 2 above)
Comment (corneil) [reply spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:249]: I have restored the logic on when we decided to remove headers. It uses extra['custom-registry']

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java (hunk 1 above)
Comment (corneil) [reply spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:230]: Fixed creation of HttpClient

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java [hunk 3]
```diff
@@ -153,67 +150,104 @@ public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolea
...
 		// Set the HTTP proxy if configured.
 		if (withHttpProxy) {
 			if (!properties.getHttpProxy().isEnabled()) {
 				throw new ContainerRegistryException("Registry Configuration uses a HttpProxy but non is configured!");
 			}
-			HttpHost proxy = new HttpHost(properties.getHttpProxy().getHost(), properties.getHttpProxy().getPort());
-			clientBuilder.setProxy(proxy);
+			ProxyProvider.Builder builder = ProxyProvider.builder().type(ProxyProvider.Proxy.HTTP).host(properties.getHttpProxy().getHost()).port(properties.getHttpProxy().getPort());
+			client.proxy(typeSpec -> builder.build());
 		}
```
Comment (onobc) [inline spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:248]: Breaking the lines on the fluent builder makes this more readable I think. You like?

```suggestion
			ProxyProvider.Builder builder = ProxyProvider.builder().type(ProxyProvider.Proxy.HTTP)
					.host(properties.getHttpProxy().getHost())
					.port(properties.getHttpProxy().getPort());
```

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java [hunk 4]
```diff
@@ -86,6 +77,12 @@
...
 
+	private static final String CUSTOM_REGISTRY = "custom-registry";
+	private static final String AMZ_CREDENTIAL = "X-Amz-Credential";
+	private static final String AUTHORIZATION_HEADER = "Authorization";
+	private static final String AZURECR_URI_SUFFIX = "azurecr.io";
+	private static final String BASIC_AUTH = "Basic";
+
 	private final RestTemplateBuilder restTemplateBuilder;
 
 	private final ContainerRegistryProperties properties;
```
Comment (onobc) [inline spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:88]: **[UNRELATED]** This was already in this class, but we can 
- use Java 17 records to reduce this CacheKey bolierplate
- use ConcurrentHashMap properly w/ `computeIfAbsent`
to shrink this down into a few lines.

The code in question is here: https://github.com/spring-cloud/spring-cloud-dataflow/pull/5996/files#diff-d659fdfefb699e093be40defb784f08e85441f3a595328ee597aa899184ef6b4R97-R151

The code would look like this once refactored:
```java
	/**
	 * Depends on the disablesSslVerification and useHttpProxy a 4 different RestTemplate configurations might be
	 * used at the same time for interacting with different container registries.
	 * The cache map allows reusing the RestTemplates for given useHttpProxy and disablesSslVerification combination.
	 */
	private final ConcurrentHashMap<CacheKey, RestTemplate> restTemplateCache;

	/**
	 * Unique key to cache rest templates by for any skipSslVerification / withHttpProxy combination.
	 */
	private record CacheKey(boolean skipSslVerification, boolean withHttpProxy) {
	}

	public ContainerImageRestTemplateFactory(RestTemplateBuilder restTemplateBuilder, ContainerRegistryProperties properties) {
		this.restTemplateBuilder = restTemplateBuilder;
		this.properties = properties;
		this.restTemplateCache = new ConcurrentHashMap<>();
	}

	public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolean withHttpProxy) {
		return this.getContainerRestTemplate(skipSslVerification, withHttpProxy, Collections.emptyMap());
	}

	public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolean withHttpProxy, Map<String, String> extra) {
		var cacheKey = new CacheKey(skipSslVerification, withHttpProxy);
		try {
			return this.restTemplateCache.computeIfAbsent(cacheKey,
					(k) -> createContainerRestTemplate(k.skipSslVerification(), k.withHttpProxy(), extra));
		}
		catch (Exception ex) {
			throw new ContainerRegistryException("Failed to create REST template for (%s)".formatted(cacheKey), ex);
		}
	}
```

If you add this just do it in a separate commit. I am also glad to do it in a polish commit during merge.

File: spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2020-2020 the original author or authors.
```
Comment (onobc) [inline spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java:2]: Please update this to `2024-2024`

File: spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import org.springframework.cloud.dataflow.container.registry.ContainerRegistryProperties;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Import;
+import org.springframework.context.annotation.Primary;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Adam J. Weigold
```
Comment (onobc) [inline spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java:39]: He did not add this.

File: spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ * @author Corneil du Plessis
+ */
+public class DropAuthorizationHeaderOnContainerTestManual {
```
Comment (onobc) [inline spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java:42]: So this says its a manual test that can be run... it looks automated to me... also please add a description to tell me how to use this test to test other registries manually.

**Review by onobc — CHANGES_REQUESTED** [review]
Excellent work @corneil .

The move to Reactor Netty "is the way" and has greatly reduced the complexity of this class. I am glad you stayed firm on getting this in - worth the wait!

A few minor comments/suggestions - but after those are resolved we should be good to go as far as I'm concerned.

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java [hunk 8]
```diff
@@ -16,53 +16,44 @@
...
+ * On demand creates a cacheable {@link RestTemplate} instances for the purpose of the
+ * Container Registry access. Created RestTemplates can be configured to use Http Proxy
+ * and/or bypassing the SSL verification.
  *
- * For configuring a Http Proxy in need to:
- *  1. Add http proxy configuration using the spring.cloud.dataflow.container.httpProxy.* properties.
- *  2. For every {@link ContainerRegistryConfiguration} that has to interact via the http proxy set the use-http-proxy
- *     flag to true. For example:
- *     <code>spring.cloud.dataflow.container.registry-configurations[reg-name].use-http-proxy=ture</code>
+ * For configuring a Http Proxy in need to: 1. Add http proxy configuration using the
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:48]: Fixed

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java [hunk 9]
```diff
@@ -153,67 +150,104 @@ public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolea
...
+	 *   (https://docs.microsoft.com/en-us/azure/container-registry/container-registry-faq#authentication-information-is-not-given-in-the-correct-format-on-direct-rest-api-calls)
+	 *
+	 * Custom:
+	 *   Custom Container Registry may have same type of issues as S3 so header needs to be dropped as well.
+	 *
+	 * @author Adam J. Weigold
+	 * @author Janne Valkealahti
+	 * @author Christian Tzolov
+	 * @author Cheng Guan Poh
+	 * @author Corneil du Plessis
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:187]: Done

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java [hunk 10]
```diff
@@ -153,67 +150,104 @@ public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolea
...
+			SslContext sslContext = builder.build();
+			HttpClient client = HttpClient.create().secure(sslContextSpec -> sslContextSpec.sslContext(sslContext));
 
-					public void checkServerTrusted(java.security.cert.X509Certificate[] certs, String authType) {
-					}
+			return client.followRedirect(true, (entries, httpClientRequest) -> {
+				if (shouldRemoveAuthorization(httpClientRequest, extra)) {
+					HttpHeaders httpHeaders = httpClientRequest.requestHeaders();
+					removeAuthorization(httpHeaders);
+					removeAuthorization(entries);
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:202]: My assumption was that a next handler may or may not merge the default headers with the request headers.

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java [hunk 11]
```diff
@@ -153,67 +150,104 @@ public RestTemplate getContainerRestTemplate(boolean skipSslVerification, boolea
...
-			.setConnectionManager(new BasicHttpClientConnectionManager(connSocketFactoryLookup));
+
+	private static void removeAuthorization(HttpHeaders headers) {
+		for(Map.Entry<String,String> entry: headers.entries()) {
+			if(entry.getKey().equalsIgnoreCase(org.springframework.http.HttpHeaders.AUTHORIZATION)) {
+				headers.remove(entry.getKey());
+				break;
+			}
+		}
 	}
```
[the remark this answers is not in the document]
Comment (corneil) [reply spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:237]: Removing an entry while iterating invariable the same list invariable causes issues. I will rather do a search and remove the result.

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java (hunk 3 above)
Comment (corneil) [reply spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:248]: done

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java (hunk 4 above)
Comment (corneil) [reply spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:88]: done

File: spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java (hunk 6 above)
Comment (corneil) [reply spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java:39]: removed

File: spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java (hunk 4 above)
Comment (corneil) [reply spring-cloud-dataflow-container-registry/src/main/java/org/springframework/cloud/dataflow/container/registry/ContainerImageRestTemplateFactory.java:88]: I also decided to add `extra` to the cache key. If different values are used for extra the behaviour of the RestTemplate may not match expectation if the cached entry wasn't created with same values in `extra`

File: spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java (hunk 7 above)
Comment (corneil) [reply spring-cloud-dataflow-configuration-metadata/src/test/java/org/springframework/cloud/dataflow/container/registry/authorization/DropAuthorizationHeaderOnContainerTestManual.java:52]: I'm updating the example and docs as for docker-hub.

I did test with `spring-scdf-docker-dev-local.usw1.packages.broadcom.com`
and `p-scdf-for-kubernetes/spring-cloud-dataflow-composed-task-runner:1.6.4` 
and my broadcom okta id without `@broadcom.net` and Artifactory access token.

**Review by onobc — APPROVED** [review]
Great work @corneil - LGTM!

**Pull request merged**